raiden/network/proxies/token_network_registry.py

```diff
@@ -453,8 +505,12 @@ def _add_token(
                     "should be larger than the minimum settlement timeout."
                 )
 
-            if self.get_token_network(token_address, failed_at_blocknumber):
-                raise RaidenRecoverableError("Token already registered")
+            total_supply = token_proxy.total_supply(block_identifier=failed_at_blocknumber)
+            if not total_supply or total_supply <= 0:
```

Comment:
why would `total_supply` here be `None` or `False` to be checked this way? constructing the token proxy from the preconditions should have checked and verified that the token exists and the bytecode matches.

Reply:
since the token is an external smart contract it can do anything, in theory it could be a proxy smart contract, and the address to which the proxy points could be changed (I have not thought if the ERC20 stuff would work, but that is sort of the point here, weird stuff can happen).

Anything that can change should not be checked only in the preconditions step, it should be checked again after the transaction is sent. The only checks that we can have exclusively in the preconditions is for our own smart contracts, which we validate the byte code,and we know that they will never change (and the pre condition check is done against a confirmed block ... but still, even in this case the precondition check could be skipped because of pruned blocks, and then we need the checks anyways).